In [1]:
import pandas as pd
melb_df = pd.read_csv('data/melb_data_fe.csv')
melb_df.head()

Suburb  Rooms   Type      Price Method SellerG        Date  Distance  \
0  Abbotsford      2  house  1480000.0      S  Biggin  2016-03-12       2.5   
1  Abbotsford      2  house  1035000.0      S  Biggin  2016-04-02       2.5   
2  Abbotsford      3  house  1465000.0     SP  Biggin  2017-04-03       2.5   
3  Abbotsford      3  house   850000.0     PI  Biggin  2017-04-03       2.5   
4  Abbotsford      4  house  1600000.0     VB  Nelson  2016-04-06       2.5   

   Postcode  Bedroom  ...  Longtitude             Regionname  Propertycount  \
0      3067        2  ...    144.9984  Northern Metropolitan           4019   
1      3067        2  ...    144.9934  Northern Metropolitan           4019   
2      3067        3  ...    144.9944  Northern Metropolitan           4019   
3      3067        3  ...    144.9969  Northern Metropolitan           4019   
4      3067        3  ...    144.9941  Northern Metropolitan           4019   

   MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding WeekdaySale  StreetType  \
0            25.20 -0.231707          3           46           5          St   
1            15.80 -0.327660          4          116           5          St   
2            18.75  0.056338          4          117           0          St   
3            15.75  0.145455          4           47           0       other   
4            17.75  0.083969          4            2           2          St   

   Weekend  
0        1  
1        1  
2        0  
3        0  
4        0  

[5 rows x 26 columns]

In [2]:
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Suburb           13580 non-null  object 
 1   Rooms            13580 non-null  int64  
 2   Type             13580 non-null  object 
 3   Price            13580 non-null  float64
 4   Method           13580 non-null  object 
 5   SellerG          13580 non-null  object 
 6   Date             13580 non-null  object 
 7   Distance         13580 non-null  float64
 8   Postcode         13580 non-null  int64  
 9   Bedroom          13580 non-null  int64  
 10  Bathroom         13580 non-null  int64  
 11  Car              13580 non-null  int64  
 12  Landsize         13580 non-null  float64
 13  BuildingArea     13580 non-null  float64
 14  CouncilArea      12211 non-null  object 
 15  Lattitude        13580 non-null  float64
 16  Longtitude       13580 non-null  float64
 17  Regionname  

None

----

###  Задание 1.1 ###

Преобразовать столбец `Date` в формат `datetime` и выделить квартал (`quarter`) продажи объектов недвижимости. Найдите второй по популярности квартал продажи. В качестве ответа записать число объектов, проданных в этом квартале.

***Ответ - 4359***

In [3]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'])
quarters = melb_df['Date'].dt.quarter
quarters.value_counts()

3    4873
2    4359
4    2329
1    2019
Name: Date, dtype: int64

----

### Задание 1.2 ###

Преобразовать все столбцы, в которых меньше 150 уникальных значений, в тип данных `category`, исключив из преобразования столбцы **`Date`**, **`Rooms`**, **`Bedroom`**, **`Bathroom`**, **`Car`**.

В качестве ответа записать результирующее количество столбцов, которые имеют тип данных `category`.

***Ответ - 10***

In [4]:
unique_list = []
for col in melb_df.columns:
    item = (col, melb_df[col].nunique(), melb_df[col].dtypes)
    unique_list.append(item)
unique_counts = pd.DataFrame(
    unique_list,
    columns=['Column_Name', 'Num_Unique', 'Type']
).sort_values(by='Num_Unique', ignore_index=True)
display(unique_counts)

Column_Name  Num_Unique            Type
0           Weekend           2           int64
1              Type           3          object
2            Method           5          object
3       WeekdaySale           7           int64
4        Regionname           8          object
5             Rooms           9           int64
6          Bathroom           9           int64
7               Car          11           int64
8        StreetType          11          object
9         MonthSale          12           int64
10          Bedroom          12           int64
11      CouncilArea          33          object
12          SellerG          50          object
13             Date          58  datetime64[ns]
14           Suburb         120          object
15      AgeBuilding         152           int64
16         Postcode         198           int64
17         Distance         202         float64
18    Propertycount         311           int64
19     BuildingArea         602         float64
20  MeanRoomsSquare        1275         float64
21         Landsize        1448         float64
22            Price        2204         float64
23        Lattitude        6503         float64
24        AreaRatio        6521         float64
25       Longtitude        7063         float64

In [5]:
cols_to_exclude = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car']
max_unique_count = 150
for col in melb_df.columns:
    if melb_df[col].nunique() < max_unique_count and col not in cols_to_exclude:
        melb_df[col] = melb_df[col].astype('category')
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Suburb           13580 non-null  category      
 1   Rooms            13580 non-null  int64         
 2   Type             13580 non-null  category      
 3   Price            13580 non-null  float64       
 4   Method           13580 non-null  category      
 5   SellerG          13580 non-null  category      
 6   Date             13580 non-null  datetime64[ns]
 7   Distance         13580 non-null  float64       
 8   Postcode         13580 non-null  int64         
 9   Bedroom          13580 non-null  int64         
 10  Bathroom         13580 non-null  int64         
 11  Car              13580 non-null  int64         
 12  Landsize         13580 non-null  float64       
 13  BuildingArea     13580 non-null  float64       
 14  CouncilArea      12211 non-null  categ

None

# Сортировка данных в DataFrame #

## Метод sort_values() ##

### Сортировка по значениям одного столбца ###

In [6]:
melb_df.sort_values(by='Price').head(10)

Suburb  Rooms   Type     Price Method        SellerG  \
2652        Footscray      1   unit   85000.0     PI        Burnham   
1805            other      4  house  131000.0     PI          other   
7303           Albion      1   unit  145000.0     PI         Biggin   
1927           Coburg      4  house  145000.0     PI         Jellis   
7940         Hawthorn      1   unit  160000.0     VB            HAR   
12666       Brunswick      1   unit  170000.0     VB         Nelson   
8811        Footscray      1   unit  170000.0     PI        Burnham   
8504   West Footscray      1   unit  185000.0     PI            Jas   
7293           Albion      1   unit  185000.0      S  hockingstuart   
7305           Albion      2   unit  190000.0     SP        Burnham   

            Date  Distance  Postcode  Bedroom  ...  Longtitude  \
2652  2016-03-09       6.4      3011        1  ...   144.89000   
1805  2017-02-25       8.9      3162        4  ...   145.02420   
7303  2016-05-28      13.9      3020        2  ...   144.82660   
1927  2016-04-06       7.8      3058        3  ...   144.96580   
7940  2017-08-04       4.6      3122        1  ...   145.03730   
12666 2017-09-16       5.2      3056        1  ...   144.95188   
8811  2017-01-07       5.1      3011        1  ...   144.89587   
8504  2017-04-29       8.2      3012        1  ...   144.86720   
7293  2016-08-22      13.9      3020        1  ...   144.82720   
7305  2016-07-30      13.9      3020        2  ...   144.82390   

                  Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
2652    Western Metropolitan           7570        42.000000  1.000000   
1805   Southern Metropolitan           2379        17.222222 -0.525994   
7303    Western Metropolitan           2185        31.500000  0.555556   
1927   Northern Metropolitan          11204        20.500000 -0.531429   
7940   Southern Metropolitan          11308        42.000000 -0.437500   
12666  Northern Metropolitan          11918        42.000000 -0.816860   
8811    Western Metropolitan           7570         8.666667 -0.071429   
8504    Western Metropolitan           5058        42.000000  1.000000   
7293    Western Metropolitan           2185        14.333333 -0.964017   
7305    Western Metropolitan           2185        25.200000  1.000000   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
2652           3            9           2          St        0  
1805           2           97           5          St        1  
7303           5           46           5          St        1  
1927           4          106           2          Rd        0  
7940           8            8           4          St        0  
12666          9           47           5          St        1  
8811           1            4           5          St        1  
8504           4           47           5          St        1  
7293           8           41           0          Rd        0  
7305           7           46           5          St        1  

[10 rows x 26 columns]

In [7]:
melb_df.sort_values(by='Date', ascending=False)

Suburb  Rooms   Type      Price Method    SellerG       Date  \
11144     Northcote      4  house  1955000.0     SP    McGrath 2017-12-08   
11217  Surrey Hills      3  house  1775000.0     PI     Jellis 2017-12-08   
11206      St Kilda      4  house  1600000.0     VB       Gary 2017-12-08   
11207    Strathmore      7  house  2000000.0     SP     Nelson 2017-12-08   
11208    Strathmore      4  house  1610000.0      S      other 2017-12-08   
...             ...    ...    ...        ...    ...        ...        ...   
1555     Camberwell      4  house  2650000.0      S     Jellis 2016-03-09   
4510       Oak Park      3  house   892000.0      S     Nelson 2016-03-09   
4388      Northcote      3  house  1200000.0      S     Nelson 2016-03-09   
6184   Surrey Hills      3  house  1205000.0      S  Fletchers 2016-01-28   
6183   Surrey Hills      2   unit   813000.0      S  Fletchers 2016-01-28   

       Distance  Postcode  Bedroom  ...  Longtitude             Regionname  \
11144       5.3      3070        4  ...   144.99375  Northern Metropolitan   
11217      10.2      3127        3  ...   145.08968  Southern Metropolitan   
11206       5.0      3182        4  ...   144.98324  Southern Metropolitan   
11207       8.2      3041        7  ...   144.90587   Western Metropolitan   
11208       8.2      3041        4  ...   144.92243   Western Metropolitan   
...         ...       ...      ...  ...         ...                    ...   
1555        7.8      3124        4  ...   145.06860  Southern Metropolitan   
4510       11.5      3046        3  ...   144.91740  Northern Metropolitan   
4388        5.5      3070        3  ...   144.99090  Northern Metropolitan   
6184       11.2      3127        3  ...   145.10060  Southern Metropolitan   
6183       11.2      3127        2  ...   145.10230  Southern Metropolitan   

       Propertycount  MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding  \
11144          11364        12.600000  0.923664         12           47   
11217           5457        15.750000 -0.721854         12           47   
11206          13240        16.000000 -0.586028         12          107   
11207           3284        20.882353 -0.262721         12           18   
11208           3284        24.909091 -0.517181         12           62   
...              ...              ...       ...        ...          ...   
1555            8920        18.900000 -0.550535          3          126   
4510            2651        18.000000 -0.714609          3           46   
4388           11364        17.857143 -0.264706          3           96   
6184            5457        18.000000 -0.590909          1           46   
6183            5457        25.200000  0.076923          1           46   

      WeekdaySale  StreetType  Weekend  
11144           4          St        0  
11217           4          Rd        0  
11206           4          St        0  
11207           4          Rd        0  
11208           4          St        0  
...           ...         ...      ...  
1555            2          Rd        0  
4510            2          St        0  
4388            2          St        0  
6184            3          Rd        0  
6183            3          Rd        0  

[13580 rows x 26 columns]

### Сортировка по значениям нескольких столбцов ###

In [8]:
melb_df.sort_values(by=['Distance', 'Price']).loc[::10, ['Distance', 'Price']]

Distance      Price
11428       0.0   387000.0
10512       0.7   600000.0
5727        1.2   485000.0
8671        1.2   595000.0
5736        1.2   740000.0
...         ...        ...
12011      38.0   680000.0
10673      38.0   810000.0
13429      38.0  1155000.0
11102      41.0   650000.0
12728      45.9   650000.0

[1358 rows x 2 columns]

In [9]:
melb_df.sort_values(by=['Price', 'Distance']).loc[::10, ['Distance', 'Price']]

Distance      Price
2652        6.4    85000.0
7116        7.8   200000.0
5206       11.2   220000.0
7240        8.2   241000.0
2898       13.0   250000.0
...         ...        ...
13341       6.3  4060000.0
8591        1.5  4275000.0
1094       11.2  4700000.0
5801        6.1  5046000.0
251         6.3  5525000.0

[1358 rows x 2 columns]

### Комбинирование сортировки с фильтрацией ###

In [10]:
mask1 = melb_df['AreaRatio'] < -0.8
mask2 = melb_df['Type'] == 'townhouse'
mask3 = melb_df['SellerG'] == 'McGrath'
melb_df[mask1 & mask2 & mask3].sort_values(
    by=['Date', 'AreaRatio'],
    ascending=[True, False],
    ignore_index=True
).loc[:, ['Date', 'AreaRatio']]

Date  AreaRatio
0 2016-07-26  -0.974922
1 2016-09-24  -0.971831
2 2016-11-27  -0.953608
3 2016-12-11  -0.945946
4 2017-08-04  -0.947368
5 2017-08-04  -0.970874

----

###  Задание 2.2 ###

Произвести сортировку столбца `AreaRatio` по убыванию. При этом индексы полученной таблицы заменить на новые. Какое значение площади здания находится в строке 1558? Ответ округлить до целого числа.

***Ответ - 126***

In [11]:
melb_df.sort_values(by='AreaRatio', ascending=False, ignore_index=True)['BuildingArea'].iloc[1555:1560]

1555     53.3
1556     93.0
1557     81.0
1558    126.0
1559     72.0
Name: BuildingArea, dtype: float64

----

###  Задание 2.3 ###

Найти таунхаусы (`Type`) с количеством жилых комнат (`Rooms`) больше 2. Отсортировать полученную таблицу сначала по возрастанию числа комнат, а затем по убыванию средней площади комнат (`MeanRoomsSquare`). Индексы таблицы заменить на новые. Какая цена будет у объекта в строке 18? Ответ запиcать в виде целого числа.

***Ответ - 1300000***

In [12]:
melb_df[(melb_df['Rooms'] > 2) & (melb_df['Type'] == 'townhouse')].sort_values(
    by=['Rooms', 'MeanRoomsSquare'],
    ascending=[True, False],
    ignore_index=True
).loc[18, 'Price']

1300000.0

# Группировка данных в DataFrame #

## Метод groupby() ##

### Группировка данных по одному критерию с одной агрегацией ###

In [13]:
melb_df.groupby(by='Type').mean()

Rooms         Price   Distance     Postcode   Bedroom  Bathroom  \
Type                                                                            
house      3.260874  1.242665e+06  10.979479  3104.080643  3.229336  1.613822   
townhouse  2.837522  9.337351e+05   9.851346  3100.777379  2.814183  1.809695   
unit       1.963871  6.051275e+05   7.607391  3110.797481  1.966523  1.183295   

                Car    Landsize  BuildingArea  Lattitude  Longtitude  \
Type                                                                   
house      1.772674  617.181924    152.162553 -37.803795  144.994700   
townhouse  1.555655  279.606822    134.649710 -37.815782  144.996489   
unit       1.128936  477.314219    102.235863 -37.823710  144.996363   

           Propertycount  MeanRoomsSquare  AreaRatio  AgeBuilding  
Type                                                               
house        7259.025505        18.996731  -0.490031    55.669700  
townhouse    7094.459605        18.569847  -0.094916    26.690305  
unit         8199.280080        21.068242   0.319883    39.703016

In [14]:
melb_df.groupby('Type')['Price'].mean()

Type
house        1.242665e+06
townhouse    9.337351e+05
unit         6.051275e+05
Name: Price, dtype: float64

In [15]:
melb_df.groupby('Regionname')['Distance'].min().sort_values(ascending=False)

Regionname
Western Victoria              29.8
Eastern Victoria              25.2
Northern Victoria             21.8
South-Eastern Metropolitan    14.7
Eastern Metropolitan           7.8
Western Metropolitan           4.3
Southern Metropolitan          0.7
Northern Metropolitan          0.0
Name: Distance, dtype: float64

### Группировка данных по одному критерию с несколькими агрегациями ###

In [16]:
melb_df.groupby('MonthSale')['Price'].agg(
    ['count', 'mean', 'max']
).sort_values(by='count', ascending=False)

count          mean        max
MonthSale                                
8           1850  1.056371e+06  6500000.0
7           1835  9.314698e+05  9000000.0
5           1644  1.097807e+06  8000000.0
6           1469  1.068981e+06  7650000.0
3           1408  1.146762e+06  5600000.0
4           1246  1.050479e+06  5500000.0
9           1188  1.126349e+06  6400000.0
10           854  1.135970e+06  6250000.0
11           750  1.142503e+06  5050000.0
12           725  1.144737e+06  5700000.0
2            333  1.169051e+06  4735000.0
1            278  9.397921e+05  5200000.0

In [17]:
melb_df.groupby('MonthSale')['Price'].agg('describe')

count          mean            std       min       25%        50%  \
MonthSale                                                                       
1           278.0  9.397921e+05  577668.924214  170000.0  570500.0   795000.0   
2           333.0  1.169051e+06  671564.357417  131000.0  710000.0  1020000.0   
3          1408.0  1.146762e+06  709573.596867   85000.0  680000.0   945000.0   
4          1246.0  1.050479e+06  591892.902979  145000.0  655000.0   905500.0   
5          1644.0  1.097807e+06  668492.867996  145000.0  650000.0   905000.0   
6          1469.0  1.068981e+06  606010.069052  222000.0  660000.0   900000.0   
7          1835.0  9.314698e+05  537390.803161  190000.0  586750.0   800000.0   
8          1850.0  1.056371e+06  619617.476541  160000.0  635000.0   892000.0   
9          1188.0  1.126349e+06  608734.690742  170000.0  725000.0   980000.0   
10          854.0  1.135970e+06  692950.251627  250000.0  652625.0   950000.0   
11          750.0  1.142503e+06  703666.227892  240000.0  671750.0   950000.0   
12          725.0  1.144737e+06  706549.007161  250000.0  667500.0   938000.0   

                 75%        max  
MonthSale                        
1          1111250.0  5200000.0  
2          1478000.0  4735000.0  
3          1400000.0  5600000.0  
4          1298750.0  5500000.0  
5          1371250.0  8000000.0  
6          1325000.0  7650000.0  
7          1150000.0  9000000.0  
8          1310000.0  6500000.0  
9          1360000.0  6400000.0  
10         1416500.0  6250000.0  
11         1399125.0  5050000.0  
12         1412000.0  5700000.0

In [18]:
melb_df.groupby('Regionname')['SellerG'].agg(['nunique', set])

nunique  \
Regionname                            
Eastern Metropolitan             26   
Eastern Victoria                 11   
Northern Metropolitan            40   
Northern Victoria                11   
South-Eastern Metropolitan       25   
Southern Metropolitan            38   
Western Metropolitan             34   
Western Victoria                  6   

                                                                          set  
Regionname                                                                     
Eastern Metropolitan        {Marshall, Fletchers, other, Ray, Jellis, Harc...  
Eastern Victoria            {Ray, Harcourts, Fletchers, McGrath, other, Ba...  
Northern Metropolitan       {Cayzer, Marshall, Collins, Fletchers, other, ...  
Northern Victoria           {Ray, Raine, McGrath, YPA, other, Barry, hocki...  
South-Eastern Metropolitan  {Fletchers, other, Eview, Ray, Jellis, Harcour...  
Southern Metropolitan       {Cayzer, Marshall, Collins, Fletchers, other, ...  
Western Metropolitan        {other, Alexkarbon, Jellis, Ray, Harcourts, Ra...  
Western Victoria                 {Ray, Raine, YPA, other, hockingstuart, HAR}

----

###  Задание 3.1 ###

Сгруппировать данные по признаку количества комнат и найти среднюю цену объектов недвижимости в каждой группе. В качестве ответа записать количество комнат, для которых средняя цена наибольшая.

***Ответ - 7***

In [19]:
display(melb_df.groupby('Rooms')['Price'].mean().max())
melb_df.groupby('Rooms')['Price'].mean()

1920700.0

Rooms
1     4.338245e+05
2     7.750812e+05
3     1.076081e+06
4     1.445282e+06
5     1.870260e+06
6     1.849366e+06
7     1.920700e+06
8     1.602750e+06
10    9.000000e+05
Name: Price, dtype: float64

----

###  Задание 3.2 ###

Какой регион имеет наименьшее стандартное отклонение по географической широте (Lattitude)?
В качестве ответа записать название этого региона.

***Ответ - Western Victoria***

In [20]:
display(melb_df.groupby('Regionname')['Lattitude'].std().min())
melb_df.groupby('Regionname')['Lattitude'].std()

0.011579033453868534

Regionname
Eastern Metropolitan          0.047890
Eastern Victoria              0.147067
Northern Metropolitan         0.049639
Northern Victoria             0.084455
South-Eastern Metropolitan    0.073411
Southern Metropolitan         0.043080
Western Metropolitan          0.051251
Western Victoria              0.011579
Name: Lattitude, dtype: float64

----

###  Задание 3.3 ###

Какая риелторская компания (`SellerG`) имеет наименьшую общую выручку за период с 1 мая по 1 сентября (включительно) 2017 года?

Для ответа на этот вопрос рассчитать сумму продаж (`Price`) каждой компании в заданный период.

Не забыть перевести даты в формат `datetime`.

***Ответ - LITTLE***

In [21]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)
begin_mask = melb_df['Date'] >= '2017-05-01'
end_mask = melb_df['Date'] <= '2017-09-01'
display(melb_df[begin_mask & end_mask].groupby('SellerG')['Price'].sum().sort_values())

SellerG
LITTLE             2742000.0
Cayzer             4439000.0
Burnham            4550500.0
Moonee             7328000.0
Thomson            8332000.0
Bells              8656000.0
Alexkarbon        10985000.0
McDonald          14637500.0
Rendina           15422276.0
Nick              16890000.0
Douglas           18341000.0
Buckingham        19033000.0
C21               19515000.0
Eview             19791500.0
Collins           20217000.0
Philip            22051800.0
Chisholm          23225000.0
Williams          23297000.0
Love              23365500.0
Purplebricks      23401000.0
O'Brien           23855508.0
HAR               25568000.0
Village           26473000.0
RW                29261000.0
Raine             30687700.0
Stockdale         35409800.0
Sweeney           36882750.0
Gary              39138400.0
Hodges            43231000.0
YPA               46354350.0
Miles             47582000.0
Kay               48569500.0
RT                50498000.0
Brad              55955000.0
Jas   

# Сводные таблицы #

## Метод groupby как способ построения сводных таблиц ##

In [22]:
melb_df.groupby('Rooms')[['Price', 'BuildingArea']].median()

Price  BuildingArea
Rooms                         
1       385000.0         107.0
2       690000.0         126.0
3       950000.0         126.0
4      1285000.0         142.0
5      1660000.0         176.0
6      1800000.0         126.0
7      1496000.0         216.5
8      1515000.0         126.0
10      900000.0         126.0

In [23]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean()

Rooms  Type     
1      house        8.668655e+05
       townhouse    5.927045e+05
       unit         3.899289e+05
2      house        1.017238e+06
       townhouse    7.101585e+05
       unit         6.104905e+05
3      house        1.109233e+06
       townhouse    9.847087e+05
       unit         8.505963e+05
4      house        1.462283e+06
       townhouse    1.217092e+06
       unit         1.037476e+06
5      house        1.877327e+06
       townhouse    1.035000e+06
       unit                  NaN
6      house        1.869508e+06
       townhouse             NaN
       unit         5.200000e+05
7      house        1.920700e+06
       townhouse             NaN
       unit                  NaN
8      house        1.510286e+06
       townhouse             NaN
       unit         2.250000e+06
10     house        9.000000e+05
       townhouse             NaN
       unit                  NaN
Name: Price, dtype: float64

In [24]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean().unstack()

Type          house     townhouse          unit
Rooms                                          
1      8.668655e+05  5.927045e+05  3.899289e+05
2      1.017238e+06  7.101585e+05  6.104905e+05
3      1.109233e+06  9.847087e+05  8.505963e+05
4      1.462283e+06  1.217092e+06  1.037476e+06
5      1.877327e+06  1.035000e+06           NaN
6      1.869508e+06           NaN  5.200000e+05
7      1.920700e+06           NaN           NaN
8      1.510286e+06           NaN  2.250000e+06
10     9.000000e+05           NaN           NaN

## Метод pivot_table для построения сводных таблиц ##

In [25]:
melb_df.pivot_table(
    values='Price',
    index='Rooms',
    columns='Type',
    fill_value=0
).round()

Type       house  townhouse       unit
Rooms                                 
1       866866.0   592705.0   389929.0
2      1017238.0   710158.0   610491.0
3      1109233.0   984709.0   850596.0
4      1462283.0  1217092.0  1037476.0
5      1877327.0  1035000.0        0.0
6      1869508.0        0.0   520000.0
7      1920700.0        0.0        0.0
8      1510286.0        0.0  2250000.0
10      900000.0        0.0        0.0

In [26]:
melb_df.pivot_table(
    values='Price',
    index='Regionname',
    columns='Weekend',
    aggfunc='count'
)

Weekend                        0     1
Regionname                            
Eastern Metropolitan         447  1024
Eastern Victoria              13    40
Northern Metropolitan       1258  2632
Northern Victoria             11    30
South-Eastern Metropolitan   123   327
Southern Metropolitan       1534  3161
Western Metropolitan         960  1988
Western Victoria               8    24

In [27]:
melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)

median                        mean              \
Type                        house townhouse unit        house   townhouse   
Regionname                                                                  
Eastern Metropolitan        674.0     233.5  203   717.422847  269.440678   
Eastern Victoria            843.0       0.0  230  3108.960000    0.000000   
Northern Metropolitan       459.5     134.0    0   619.249092  317.325733   
Northern Victoria           724.0       0.0    0  3355.463415    0.000000   
South-Eastern Metropolitan  630.5     240.0  199   664.306701  212.160000   
Southern Metropolitan       586.0     246.0    0   569.643881  278.858824   
Western Metropolitan        531.0     198.0   62   507.883406  244.560669   
Western Victoria            599.5       0.0    0   655.500000    0.000000   

                                        
Type                              unit  
Regionname                              
Eastern Metropolitan        330.444444  
Eastern Victoria            295.333333  
Northern Metropolitan       495.026538  
Northern Victoria             0.000000  
South-Eastern Metropolitan  357.864865  
Southern Metropolitan       466.380245  
Western Metropolitan        557.637232  
Western Victoria              0.000000

## Многомерные сводные таблицы ##

In [28]:
melb_df.pivot_table(
    values='Price',
    index=['Method', 'Type'],
    columns='Regionname',
    aggfunc='median',
    fill_value=0
)

Regionname        Eastern Metropolitan  Eastern Victoria  \
Method Type                                                
PI     house                   1244000            780000   
       townhouse                760000                 0   
       unit                     650000                 0   
S      house                   1127000            675000   
       townhouse                828000                 0   
       unit                     645750            492000   
SA     house                    932500            950000   
       townhouse                807500                 0   
       unit                          0                 0   
SP     house                   1050000            672500   
       townhouse                910000                 0   
       unit                     515000            400000   
VB     house                   1100000            712500   
       townhouse                892500                 0   
       unit                     500000                 0   

Regionname        Northern Metropolitan  Northern Victoria  \
Method Type                                                  
PI     house                     900000             500000   
       townhouse                 632500                  0   
       unit                      410000                  0   
S      house                     920000             555000   
       townhouse                 750000                  0   
       unit                      525500                  0   
SA     house                     817500             540000   
       townhouse                 425000                  0   
       unit                      616000                  0   
SP     house                     900000             521000   
       townhouse                 690000                  0   
       unit                      470000                  0   
VB     house                    1050000             690000   
       townhouse                 640000                  0   
       unit                      450000                  0   

Regionname        South-Eastern Metropolitan  Southern Metropolitan  \
Method Type                                                           
PI     house                          865000                1725000   
       townhouse                     1190000                1055000   
       unit                           525000                 571250   
S      house                          883300                1611000   
       townhouse                      875000                1135000   
       unit                           606000                 655000   
SA     house                          880000                1390000   
       townhouse                           0                1141000   
       unit                                0                 580000   
SP     house                          770000                1521750   
       townhouse                      800000                1162500   
       unit                           601000                 550000   
VB     house                          850000                1800000   
       townhouse                           0                1250000   
       unit                           700000                 500000   

Regionname        Western Metropolitan  Western Victoria  
Method Type                                               
PI     house                    870000            630000  
       townhouse                670000                 0  
       unit                     360000                 0  
S      house                    870000            397500  
       townhouse                729000                 0  
       unit                     489000                 0  
SA     house                    772500                 0  
       townhouse                467500                 0  
       unit                     571000                 0  
SP     house                    865000            360000  
       t

## Доступ к данным в сводной таблице ##

In [29]:
pivot = melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)
pivot.columns

MultiIndex([('median',     'house'),
            ('median', 'townhouse'),
            ('median',      'unit'),
            (  'mean',     'house'),
            (  'mean', 'townhouse'),
            (  'mean',      'unit')],
           names=[None, 'Type'])

In [30]:
display(pivot)
display(pivot['mean'])
display(pivot['mean']['unit'])

median                        mean              \
Type                        house townhouse unit        house   townhouse   
Regionname                                                                  
Eastern Metropolitan        674.0     233.5  203   717.422847  269.440678   
Eastern Victoria            843.0       0.0  230  3108.960000    0.000000   
Northern Metropolitan       459.5     134.0    0   619.249092  317.325733   
Northern Victoria           724.0       0.0    0  3355.463415    0.000000   
South-Eastern Metropolitan  630.5     240.0  199   664.306701  212.160000   
Southern Metropolitan       586.0     246.0    0   569.643881  278.858824   
Western Metropolitan        531.0     198.0   62   507.883406  244.560669   
Western Victoria            599.5       0.0    0   655.500000    0.000000   

                                        
Type                              unit  
Regionname                              
Eastern Metropolitan        330.444444  
Eastern Victoria            295.333333  
Northern Metropolitan       495.026538  
Northern Victoria             0.000000  
South-Eastern Metropolitan  357.864865  
Southern Metropolitan       466.380245  
Western Metropolitan        557.637232  
Western Victoria              0.000000

Type                              house   townhouse        unit
Regionname                                                     
Eastern Metropolitan         717.422847  269.440678  330.444444
Eastern Victoria            3108.960000    0.000000  295.333333
Northern Metropolitan        619.249092  317.325733  495.026538
Northern Victoria           3355.463415    0.000000    0.000000
South-Eastern Metropolitan   664.306701  212.160000  357.864865
Southern Metropolitan        569.643881  278.858824  466.380245
Western Metropolitan         507.883406  244.560669  557.637232
Western Victoria             655.500000    0.000000    0.000000

Regionname
Eastern Metropolitan          330.444444
Eastern Victoria              295.333333
Northern Metropolitan         495.026538
Northern Victoria               0.000000
South-Eastern Metropolitan    357.864865
Southern Metropolitan         466.380245
Western Metropolitan          557.637232
Western Victoria                0.000000
Name: unit, dtype: float64

In [31]:
mask = pivot['mean']['house'] < pivot['median']['house']
filtered_pivot = pivot[mask]
display(filtered_pivot)

print(list(filtered_pivot.index))

median                       mean              \
Type                   house townhouse unit       house   townhouse   
Regionname                                                            
Southern Metropolitan  586.0     246.0    0  569.643881  278.858824   
Western Metropolitan   531.0     198.0   62  507.883406  244.560669   

                                   
Type                         unit  
Regionname                         
Southern Metropolitan  466.380245  
Western Metropolitan   557.637232

['Southern Metropolitan', 'Western Metropolitan']


In [32]:
import numpy as np

mser = pd.Series(
    np.random.rand(8),
    index=[['white', 'white', 'white', 'blue', 'blue', 'red', 'red', 'red'],
           ['up', 'down', 'right', 'up', 'down', 'up', 'down', 'left']
          ]
)
display(mser)

white  up       0.611221
       down     0.011606
       right    0.214507
blue   up       0.263766
       down     0.710844
red    up       0.818500
       down     0.534114
       left     0.232005
dtype: float64

In [33]:
print(mser.index)

MultiIndex([('white',    'up'),
            ('white',  'down'),
            ('white', 'right'),
            ( 'blue',    'up'),
            ( 'blue',  'down'),
            (  'red',    'up'),
            (  'red',  'down'),
            (  'red',  'left')],
           )


In [34]:
mframe = pd.DataFrame(
    np.random.randn(16).reshape(4, 4),
    index=[['white', 'white', 'red', 'red'],
           ['up', 'down', 'up', 'down']],
    columns=[['pen', 'pen', 'paper', 'paper'],
             [1, 2, 1, 2]]
)
display(mframe)

pen               paper          
                   1         2         1         2
white up   -0.532155  0.078085  0.311825 -0.315176
      down -0.040420 -0.208596 -0.150475 -1.596512
red   up   -0.707143  0.377259  1.406821 -0.495502
      down -0.044027 -1.006960 -0.198403 -0.367341

----

###  Задание 4.1 ###

Какой параметр метода `pivot_table()` отвечает за признак, по которому будут рассчитаны агрегирующие функции?

***Ответ - values***

----

###  Задание 4.2 ###

Составить сводную таблицу, которая показывает зависимость медианной площади (`BuildingArea`) здания от типа объекта недвижимости (`Type`) и количества жилых комнат в доме (`Rooms`). Для какой комбинации признаков площадь здания наибольшая?

В качестве ответа записать эту комбинацию (тип здания, число комнат) через запятую, без пробелов.

***Ответ - house,7***

In [35]:
melb_df.groupby(['Type', 'Rooms'])['BuildingArea'].median().sort_values(ascending=False)

Type       Rooms
house      7        216.5
           5        177.0
unit       6        171.0
townhouse  4        159.5
           5        152.0
house      4        141.0
           1        126.0
townhouse  3        126.0
unit       4        126.0
           3        126.0
           8        126.0
house      2        126.0
           10       126.0
           8        126.0
           6        126.0
           3        126.0
townhouse  2        114.0
unit       2        110.0
townhouse  1         88.0
unit       1         69.5
townhouse  6          NaN
           7          NaN
           8          NaN
           10         NaN
unit       5          NaN
           7          NaN
           10         NaN
Name: BuildingArea, dtype: float64

In [36]:
melb_df.pivot_table(
    index=['Type', 'Rooms'],
    values='BuildingArea',
    aggfunc='median',
    fill_value=0
).sort_values(by='BuildingArea', ascending=False)

BuildingArea
Type      Rooms              
house     7             216.5
          5             177.0
unit      6             171.0
townhouse 4             159.5
          5             152.0
house     4             141.0
          1             126.0
townhouse 3             126.0
unit      4             126.0
          3             126.0
          8             126.0
house     2             126.0
          10            126.0
          8             126.0
          6             126.0
          3             126.0
townhouse 2             114.0
unit      2             110.0
townhouse 1              88.0
unit      1              69.5

----

###  Задание 4.3 ###

Составить сводную таблицу, которая показывает зависимость медианной цены объекта недвижимости (`Price`) от риелторского агентства (`SellerG`) и типа здания (`Type`).

Во вновь созданной таблице найти агентство, у которого медианная цена для зданий типа `unit` максимальна. В качестве ответа записать название этого агентства.

***Ответ - Nick***

In [37]:
melb_df.pivot_table(
    values='Price',
    columns='Type',
    index='SellerG',
    aggfunc='median'
)['unit'].sort_values(ascending=False)

SellerG
Nick             900000.0
Marshall         715000.0
Cayzer           707500.0
Kay              695000.0
Noel             693250.0
Buxton           670000.0
Fletchers        653000.0
Chisholm         640000.0
Philip           636000.0
RT               630000.0
C21              625500.0
RW               625500.0
Moonee           622000.0
O'Brien          618500.0
Hodges           605500.0
Eview            602000.0
Woodards         600000.0
Collins          592000.0
Jellis           591750.0
Purplebricks     582500.0
Williams         577000.0
hockingstuart    566000.0
Alexkarbon       560000.0
Miles            555000.0
Buckingham       553000.0
Gary             550000.0
Biggin           548000.0
Greg             547500.0
McGrath          542500.0
other            535000.0
Nelson           530000.0
HAR              518000.0
McDonald         502500.0
Barry            502000.0
Harcourts        500000.0
Ray              497500.0
Thomson          495000.0
Douglas          494500.0
Brad

# Объединение DataFrame: знакомимся с новыми данными #

----

###  Задание 5.1 ###

Значения из какого столбца таблиц `ratings1` и `ratings2` можно расшифровать с помощью таблицы `movies`?

***Ответ - movieId***

In [38]:
ratings1 = pd.read_csv('data/ratings1.csv')
ratings2 = pd.read_csv('data/ratings2.csv')
movies = pd.read_csv('data/movies.csv')
display(ratings1)
display(ratings2)
display(movies)

userId  movieId  rating
0           1        1     4.0
1           1        3     4.0
2           1        6     4.0
3           1       47     5.0
4           1       50     5.0
...       ...      ...     ...
39996     274     5582     2.5
39997     274     5585     3.5
39998     274     5588     3.5
39999     274     5618     4.0
40000     274     5621     2.0

[40001 rows x 3 columns]

userId  movieId  rating
0         274     5621     2.0
1         274     5630     3.0
2         274     5667     3.5
3         274     5679     3.5
4         274     5690     3.0
...       ...      ...     ...
60831     610   166534     4.0
60832     610   168248     5.0
60833     610   168250     5.0
60834     610   168252     5.0
60835     610   170875     3.0

[60836 rows x 3 columns]

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

----

###  Задание 5.2 ###

Сколько уникальных фильмов представлено в таблице `movies`?

***Ответ - 9742***

In [39]:
movies.nunique()

movieId    9742
title      9737
genres      951
dtype: int64

----

###  Задание 5.3 ###

Сколько уникальных пользователей в таблице `ratings1`?

***Ответ - 274***

In [40]:
ratings1.nunique()

userId      274
movieId    6219
rating       10
dtype: int64

----

###  Задание 5.4 ###

В каком году было выставлено больше всего оценок?

Для ответа на этот вопрос использовать таблицу `dates`.

***Ответ - 2000***

In [41]:
dates = pd.read_csv('data/dates.csv')
dates['date'] = pd.to_datetime(dates['date'])
display(dates.info())
years = dates['date'].dt.year
display(years.value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 1 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   date    100836 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 787.9 KB


None

2000    10061
2017     8198
2007     7114
2016     6703
2015     6616
2018     6418
1996     6040
2005     5813
2012     4656
2008     4351
2009     4158
2006     4059
2003     4014
2001     3922
2002     3478
2004     3279
1999     2439
2010     2301
1997     1916
2011     1690
2013     1664
2014     1439
1998      507
Name: date, dtype: int64

# Объединение DataFrame: concat #

In [42]:
ratings = pd.concat([ratings1, ratings2])
display(ratings)

userId  movieId  rating
0           1        1     4.0
1           1        3     4.0
2           1        6     4.0
3           1       47     5.0
4           1       50     5.0
...       ...      ...     ...
60831     610   166534     4.0
60832     610   168248     5.0
60833     610   168250     5.0
60834     610   168252     5.0
60835     610   170875     3.0

[100837 rows x 3 columns]

In [43]:
ratings = pd.concat([ratings1, ratings2], ignore_index=True)
display(ratings)

userId  movieId  rating
0            1        1     4.0
1            1        3     4.0
2            1        6     4.0
3            1       47     5.0
4            1       50     5.0
...        ...      ...     ...
100832     610   166534     4.0
100833     610   168248     5.0
100834     610   168250     5.0
100835     610   168252     5.0
100836     610   170875     3.0

[100837 rows x 3 columns]

In [46]:
print('Число строк ratings:', ratings.shape[0])
print('Число строк dates:', dates.shape[0])
print(ratings.shape[0] == dates.shape[0])

Число строк ratings: 100837
Число строк dates: 100836
False


In [47]:
display(ratings1.tail(1))
display(ratings2.head(1))

userId  movieId  rating
40000     274     5621     2.0

userId  movieId  rating
0     274     5621     2.0

In [48]:
ratings = ratings.drop_duplicates(ignore_index=True)
print('Число строк ratings:', ratings.shape[0])

Число строк ratings: 100836


In [49]:
ratings_dates = pd.concat([ratings, dates], axis=1)
display(ratings_dates.tail(10))

userId  movieId  rating                date
100826     610   162350     3.5 2017-05-03 22:19:31
100827     610   163937     3.5 2017-05-03 21:59:49
100828     610   163981     3.5 2017-05-03 22:22:35
100829     610   164179     5.0 2017-05-03 21:07:11
100830     610   166528     4.0 2017-05-04 06:29:25
100831     610   166534     4.0 2017-05-03 21:53:22
100832     610   168248     5.0 2017-05-03 22:21:31
100833     610   168250     5.0 2017-05-08 19:50:47
100834     610   168252     5.0 2017-05-03 21:19:12
100835     610   170875     3.0 2017-05-03 21:20:15

In [50]:
ratings_dates.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype         
---  ------   --------------   -----         
 0   userId   100836 non-null  int64         
 1   movieId  100836 non-null  int64         
 2   rating   100836 non-null  float64       
 3   date     100836 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(2)
memory usage: 3.1 MB


----

###  Задание 6.1 ###

Какой параметр функции `concat` позволяет управлять способом конкатенации (проводить конкатенацию по строкам или по столбцам)?

***Ответ - axis***

----

###  Задание 6.2 ###

Заданы две таблицы — `df1` и `df2`. В первой содержатся имена и фамилии сотрудников, во второй — их должности.
```
df1 = pd.DataFrame({"Name": ["Pankaj", "Lisa"], "Surname": ["Sobolev", "Krasnova"]})
df2 = pd.DataFrame({"Role": ["Admin", "Editor"]})
```
Какой из способов будет верным при объединении этих таблиц?

***Ответ - df = pd.concat([df1, df2], axis=1)***

In [51]:
df1 = pd.DataFrame({"Name": ["Pankaj", "Lisa"], "Surname": ["Sobolev", "Krasnova"]})
df2 = pd.DataFrame({"Role": ["Admin", "Editor"]})
display(df1)
display(df2)

Name   Surname
0  Pankaj   Sobolev
1    Lisa  Krasnova

Role
0   Admin
1  Editor

In [52]:
df = pd.concat([df1, df2], axis=1)
display(df)

Name   Surname    Role
0  Pankaj   Sobolev   Admin
1    Lisa  Krasnova  Editor

----

### Задание 6.3 ###

Предоставлена директория `./Root/users/`. В данной директории содержатся csv-файлы, в каждом из которых хранится информация об идентификаторах пользователей (`user_id`) и ссылки на их фотографии (`photo_url`). Каждый файл из папки `users` имеет примерно следующую структуру:
```
  user_id                         image_url
0   id001  http://example.com/img/id001.png
1   id002  http://example.com/img/id002.jpg
2   id003  http://example.com/img/id003.bmp
```
При проверке в директории может быть **сколько угодно** файлов (директория может изменяться в зависимости от устройства файловой системы).

Необходимо написать функцию `concat_user_files(path)`, параметром которой является `path` — путь до директории. Функция должна объединить информацию из предоставленных файлов в один **DataFrame** и вернуть его. 

Список названий всех файлов, находящихся в директории, можно получить с помощью функции `os.listdir(path)` из модуля `os`. Например, для директории `./Root/users/` результатом работы функции будет список:
```
print(os.listdir('./Root/users/'))
['users2.csv', 'users1.csv', 'users3.csv']
```
***Примечание.***    
Модуль `os` позволяет работать с операционной системой компьютера прямо из кода.

Отсортировать этот список, прежде чем производить объединение файлов.

Когда получится отсортированный список, останется только прочитать все csv-файлы из списка в цикле и объединить прочитанные таблицы между собой.

***Примечание.***    
Метод `os.listdir()` возвращает только список имён файлов в указанной директории, а при чтении файла необходимо указывать полный путь до него. То есть путь для чтения будет таким:
```
'./Root/users/{file_name}'
```
Не забыть обновить индексы результирующей таблицы после объединения.

Учесть, что на тестовом наборе файлов в результате объединения могут возникнуть дубликаты, от которых необходимо будет избавиться.

Например, для директории "./Root/users/" результирующая таблица должна иметь следующий вид:
```
  user_id                         image_url
0   id001  http://example.com/img/id001.png
1   id002  http://example.com/img/id002.jpg
2   id003  http://example.com/img/id003.bmp
3   id004  http://example.com/img/id004.jpg
4   id005  http://example.com/img/id005.png
5   id006  http://example.com/img/id006.png
6   id007  http://example.com/img/id007.png
7   id008  http://example.com/img/id008.png
```

In [53]:
import os

DIR = 'Task63'

def concat_users_files(path_):
    file_list = os.listdir(path=path_)
    file_list.sort()
    for index in range(len(file_list)):
        file_list[index] = path_ + '/' + file_list[index]
    df = pd.read_csv(file_list[0], sep=',')
    if len(file_list) > 1:
        for file_path in file_list[1:]:
            df_add = pd.read_csv(file_path, sep=',')
            df = pd.concat([df, df_add], ignore_index=True)
    df.drop_duplicates(inplace=True, ignore_index=True)
    return df


data = concat_users_files(DIR)
print(data)

  user_id                         image_url
0   id001  http://example.com/img/id001.png
1   id002  http://example.com/img/id002.jpg
2   id003  http://example.com/img/id003.bmp
3   id004  http://example.com/img/id004.jpg
4   id005  http://example.com/img/id005.png
5   id006  http://example.com/img/id006.png
6   id007  http://example.com/img/id007.png
7   id008  http://example.com/img/id008.png


# Объединение DataFrame: join, merge #

## Метод объединения join ##

In [54]:
joined_false = ratings_dates.join(
    movies,
    rsuffix='_right',
    how='left'
)
display(joined_false)

userId  movieId  rating                date  movieId_right  \
0            1        1     4.0 2000-07-30 18:45:03            1.0   
1            1        3     4.0 2000-07-30 18:20:47            2.0   
2            1        6     4.0 2000-07-30 18:37:04            3.0   
3            1       47     5.0 2000-07-30 19:03:35            4.0   
4            1       50     5.0 2000-07-30 18:48:51            5.0   
...        ...      ...     ...                 ...            ...   
100831     610   166534     4.0 2017-05-03 21:53:22            NaN   
100832     610   168248     5.0 2017-05-03 22:21:31            NaN   
100833     610   168250     5.0 2017-05-08 19:50:47            NaN   
100834     610   168252     5.0 2017-05-03 21:19:12            NaN   
100835     610   170875     3.0 2017-05-03 21:20:15            NaN   

                                     title  \
0                         Toy Story (1995)   
1                           Jumanji (1995)   
2                  Grumpier Old Men (1995)   
3                 Waiting to Exhale (1995)   
4       Father of the Bride Part II (1995)   
...                                    ...   
100831                                 NaN   
100832                                 NaN   
100833                                 NaN   
100834                                 NaN   
100835                                 NaN   

                                             genres  
0       Adventure|Animation|Children|Comedy|Fantasy  
1                        Adventure|Children|Fantasy  
2                                    Comedy|Romance  
3                              Comedy|Drama|Romance  
4                                            Comedy  
...                                             ...  
100831                                          NaN  
100832                                          NaN  
100833                                          NaN  
100834                                          NaN  
100835                                          NaN  

[100836 rows x 7 columns]

In [55]:
joined = ratings_dates.join(
    movies.set_index('movieId'),
    on='movieId',
    how='left'
)
display(joined)

userId  movieId  rating                date  \
0            1        1     4.0 2000-07-30 18:45:03   
1            1        3     4.0 2000-07-30 18:20:47   
2            1        6     4.0 2000-07-30 18:37:04   
3            1       47     5.0 2000-07-30 19:03:35   
4            1       50     5.0 2000-07-30 18:48:51   
...        ...      ...     ...                 ...   
100831     610   166534     4.0 2017-05-03 21:53:22   
100832     610   168248     5.0 2017-05-03 22:21:31   
100833     610   168250     5.0 2017-05-08 19:50:47   
100834     610   168252     5.0 2017-05-03 21:19:12   
100835     610   170875     3.0 2017-05-03 21:20:15   

                                 title  \
0                     Toy Story (1995)   
1              Grumpier Old Men (1995)   
2                          Heat (1995)   
3          Seven (a.k.a. Se7en) (1995)   
4           Usual Suspects, The (1995)   
...                                ...   
100831                    Split (2017)   
100832   John Wick: Chapter Two (2017)   
100833                  Get Out (2017)   
100834                    Logan (2017)   
100835  The Fate of the Furious (2017)   

                                             genres  
0       Adventure|Animation|Children|Comedy|Fantasy  
1                                    Comedy|Romance  
2                             Action|Crime|Thriller  
3                                  Mystery|Thriller  
4                            Crime|Mystery|Thriller  
...                                             ...  
100831                        Drama|Horror|Thriller  
100832                        Action|Crime|Thriller  
100833                                       Horror  
100834                                Action|Sci-Fi  
100835                  Action|Crime|Drama|Thriller  

[100836 rows x 6 columns]

## Метод объединения merge ##

In [56]:
merged = ratings_dates.merge(
    movies,
    on='movieId',
    how='left'
)
display(merged)

userId  movieId  rating                date  \
0            1        1     4.0 2000-07-30 18:45:03   
1            1        3     4.0 2000-07-30 18:20:47   
2            1        6     4.0 2000-07-30 18:37:04   
3            1       47     5.0 2000-07-30 19:03:35   
4            1       50     5.0 2000-07-30 18:48:51   
...        ...      ...     ...                 ...   
100831     610   166534     4.0 2017-05-03 21:53:22   
100832     610   168248     5.0 2017-05-03 22:21:31   
100833     610   168250     5.0 2017-05-08 19:50:47   
100834     610   168252     5.0 2017-05-03 21:19:12   
100835     610   170875     3.0 2017-05-03 21:20:15   

                                 title  \
0                     Toy Story (1995)   
1              Grumpier Old Men (1995)   
2                          Heat (1995)   
3          Seven (a.k.a. Se7en) (1995)   
4           Usual Suspects, The (1995)   
...                                ...   
100831                    Split (2017)   
100832   John Wick: Chapter Two (2017)   
100833                  Get Out (2017)   
100834                    Logan (2017)   
100835  The Fate of the Furious (2017)   

                                             genres  
0       Adventure|Animation|Children|Comedy|Fantasy  
1                                    Comedy|Romance  
2                             Action|Crime|Thriller  
3                                  Mystery|Thriller  
4                            Crime|Mystery|Thriller  
...                                             ...  
100831                        Drama|Horror|Thriller  
100832                        Action|Crime|Thriller  
100833                                       Horror  
100834                                Action|Sci-Fi  
100835                  Action|Crime|Drama|Thriller  

[100836 rows x 6 columns]

In [57]:
print('Число строк rating_dates:', ratings_dates.shape[0])
print('Число строк merged:', merged.shape[0])
print(ratings_dates.shape[0] == merged.shape[0])

Число строк rating_dates: 100836
Число строк merged: 100836
True


## Особенности использования merge() ##

In [58]:
merged2 = ratings_dates.merge(
    movies,
    on='movieId',
    how='outer'
)
print('Число строк merged2:', merged2.shape[0])
display(merged2)

Число строк merged2: 100854


userId  movieId  rating                date  \
0          1.0        1     4.0 2000-07-30 18:45:03   
1          5.0        1     4.0 1996-11-08 06:36:02   
2          7.0        1     4.5 2005-01-25 06:52:26   
3         15.0        1     2.5 2017-11-13 12:59:30   
4         17.0        1     4.5 2011-05-18 05:28:03   
...        ...      ...     ...                 ...   
100849     NaN    30892     NaN                 NaT   
100850     NaN    32160     NaN                 NaT   
100851     NaN    32371     NaN                 NaT   
100852     NaN    34482     NaN                 NaT   
100853     NaN    85565     NaN                 NaT   

                                     title  \
0                         Toy Story (1995)   
1                         Toy Story (1995)   
2                         Toy Story (1995)   
3                         Toy Story (1995)   
4                         Toy Story (1995)   
...                                    ...   
100849  In the Realms of the Unreal (2004)   
100850            Twentieth Century (1934)   
100851           Call Northside 777 (1948)   
100852        Browning Version, The (1951)   
100853                  Chalet Girl (2011)   

                                             genres  
0       Adventure|Animation|Children|Comedy|Fantasy  
1       Adventure|Animation|Children|Comedy|Fantasy  
2       Adventure|Animation|Children|Comedy|Fantasy  
3       Adventure|Animation|Children|Comedy|Fantasy  
4       Adventure|Animation|Children|Comedy|Fantasy  
...                                             ...  
100849                        Animation|Documentary  
100850                                       Comedy  
100851                        Crime|Drama|Film-Noir  
100852                                        Drama  
100853                               Comedy|Romance  

[100854 rows x 6 columns]

In [59]:
merge_ratings = ratings1.merge(ratings2, how='outer')
print('Число строк merge_ratings', merge_ratings.shape[0])
display(merge_ratings)

Число строк merge_ratings 100836


userId  movieId  rating
0            1        1     4.0
1            1        3     4.0
2            1        6     4.0
3            1       47     5.0
4            1       50     5.0
...        ...      ...     ...
100831     610   166534     4.0
100832     610   168248     5.0
100833     610   168250     5.0
100834     610   168252     5.0
100835     610   170875     3.0

[100836 rows x 3 columns]

----

###  Задание 7.1 ###

Какой параметр методов `join()` и `merge()` отвечает за используемый тип объединения?

***Ответ - how***

----

###  Задание 7.2 ###

В документации по методу `merge()` найти параметры, которые позволяют управлять объединением таблиц по индексам.

***Ответ - left_index, right_index***

----

###  Задание 7.3 ###

Дано две исходных таблицы `data_1` и `data_2`.

Какой из перечисленных вариантов кода позволяет получить представленный результат?

***Ответ - data_1.join(data_2, how='inner')***

In [60]:
data_1 = pd.DataFrame({'Value': [100, 45, 80],
                       'Group': [1, 4, 5]},
                      index = ['I0', 'I1', 'I2']
                     )
data_2 = pd.DataFrame({'Company': ['Google', 'Amazon', 'Facebook'],
                       'Add': ['S0', 'S1', 'S7']},
                      index = ['I0', 'I1', 'I3']
                     )
display(data_1)
display(data_2)

Value  Group
I0    100      1
I1     45      4
I2     80      5

Company Add
I0    Google  S0
I1    Amazon  S1
I3  Facebook  S7

In [61]:
data_1.join(data_2, how='inner')

Value  Group Company Add
I0    100      1  Google  S0
I1     45      4  Amazon  S1

----

###  Задание 7.4 ###

Даны две исходные таблицы `a` и `b`.

Какой из перечисленных вариантов кода позволяет получить представленный ниже результат?

***Ответ - a.merge(b, how='right', on='C')***

In [62]:
a = pd.DataFrame({'A': ['a', 'b', 'c'], 'B': [103, 214, 124], 'C': [1, 4, 2]})
b = pd.DataFrame({'V': ['d', 'b', 'c'], 'U': [1393.7, 9382.2, 1904.5], 'C': [1, 3, 2]})
display(a)
display(b)

A    B  C
0  a  103  1
1  b  214  4
2  c  124  2

V       U  C
0  d  1393.7  1
1  b  9382.2  3
2  c  1904.5  2

In [63]:
a.merge(b, how='right', on='C')

A      B  C  V       U
0    a  103.0  1  d  1393.7
1  NaN    NaN  3  b  9382.2
2    c  124.0  2  c  1904.5

----

### Задание 7.5 ###

Даны две таблицы: `items_df`, в которой содержится информация о наличии товаров на складе, и `purchase_df` — с данными о покупках товаров.

Информация в таблицах представлена в виде следующих столбцов:
- `item_id` — идентификатор модели;
- `vendor` — производитель модели;
- `stock_count` — имеющееся на складе количество данных моделей (в штуках);
- `purchase_id` — идентификатор покупки;
- `price` — стоимость модели в покупке.

1. Сформировать **DataFrame** `merged`, в котором в результате объединения `purchase_df` и `items_df` останутся модели, которые учтены на складе и имели продажи.
2. Из таблицы `merged` найти суммарную выручку, которую можно было бы получить от продажи всех товаров, которые учтены на складе и имели продажи. Результат занести в переменную `income`.


In [64]:
items_df = pd.DataFrame({
    'item_id': [417283, 849734, 132223, 573943, 19475, 3294095, 382043, 302948, 100132, 312394], 
    'vendor': ['Samsung', 'LG', 'Apple', 'Apple', 'LG', 'Apple', 'Samsung', 'Samsung', 'LG', 'ZTE'],
    'stock_count': [54, 33, 122, 18, 102, 43, 77, 143, 60, 19]
})
purchase_df = pd.DataFrame({
    'purchase_id': [101, 101, 101, 112, 121, 145, 145, 145, 145, 221],
    'item_id': [417283, 849734, 132223, 573943, 19475, 3294095, 382043, 302948, 103845, 100132], 
    'price': [13900, 5330, 38200, 49990, 9890, 33000, 67500, 34500, 89900, 11400]
})
display(items_df)
display(purchase_df)

item_id   vendor  stock_count
0   417283  Samsung           54
1   849734       LG           33
2   132223    Apple          122
3   573943    Apple           18
4    19475       LG          102
5  3294095    Apple           43
6   382043  Samsung           77
7   302948  Samsung          143
8   100132       LG           60
9   312394      ZTE           19

purchase_id  item_id  price
0          101   417283  13900
1          101   849734   5330
2          101   132223  38200
3          112   573943  49990
4          121    19475   9890
5          145  3294095  33000
6          145   382043  67500
7          145   302948  34500
8          145   103845  89900
9          221   100132  11400

In [65]:
merged = items_df.merge(purchase_df, on='item_id', how='inner')
display(merged)

item_id   vendor  stock_count  purchase_id  price
0   417283  Samsung           54          101  13900
1   849734       LG           33          101   5330
2   132223    Apple          122          101  38200
3   573943    Apple           18          112  49990
4    19475       LG          102          121   9890
5  3294095    Apple           43          145  33000
6   382043  Samsung           77          145  67500
7   302948  Samsung          143          145  34500
8   100132       LG           60          221  11400

In [66]:
merged['total'] = merged['price'] * merged['stock_count']
display(merged)
income = merged['total'].sum()
print(income)

item_id   vendor  stock_count  purchase_id  price    total
0   417283  Samsung           54          101  13900   750600
1   849734       LG           33          101   5330   175890
2   132223    Apple          122          101  38200  4660400
3   573943    Apple           18          112  49990   899820
4    19475       LG          102          121   9890  1008780
5  3294095    Apple           43          145  33000  1419000
6   382043  Samsung           77          145  67500  5197500
7   302948  Samsung          143          145  34500  4933500
8   100132       LG           60          221  11400   684000

19729490


# Финальные задания #

In [67]:
import re

def get_year_release(arg):
    candidates = re.findall(r'\(\d{4}\)', arg)
    if len(candidates) > 0:
        year = candidates[0].replace('(', '').replace(')', '')
        return int(year)
    else:
        return None

In [68]:
df_sample = pd.read_csv('data/ratings_movies.csv')
display(df_sample)

Unnamed: 0  userId  movieId  rating                 date  \
0                0       1        1     4.0  2000-07-30 18:45:03   
1                1       1        3     4.0  2000-07-30 18:20:47   
2                2       1        6     4.0  2000-07-30 18:37:04   
3                3       1       47     5.0  2000-07-30 19:03:35   
4                4       1       50     5.0  2000-07-30 18:48:51   
...            ...     ...      ...     ...                  ...   
100831      100831     610   166534     4.0  2017-05-03 21:53:22   
100832      100832     610   168248     5.0  2017-05-03 22:21:31   
100833      100833     610   168250     5.0  2017-05-08 19:50:47   
100834      100834     610   168252     5.0  2017-05-03 21:19:12   
100835      100835     610   170875     3.0  2017-05-03 21:20:15   

                                 title  \
0                     Toy Story (1995)   
1              Grumpier Old Men (1995)   
2                          Heat (1995)   
3          Seven (a.k.a. Se7en) (1995)   
4           Usual Suspects, The (1995)   
...                                ...   
100831                    Split (2017)   
100832   John Wick: Chapter Two (2017)   
100833                  Get Out (2017)   
100834                    Logan (2017)   
100835  The Fate of the Furious (2017)   

                                             genres  
0       Adventure|Animation|Children|Comedy|Fantasy  
1                                    Comedy|Romance  
2                             Action|Crime|Thriller  
3                                  Mystery|Thriller  
4                            Crime|Mystery|Thriller  
...                                             ...  
100831                        Drama|Horror|Thriller  
100832                        Action|Crime|Thriller  
100833                                       Horror  
100834                                Action|Sci-Fi  
100835                  Action|Crime|Drama|Thriller  

[100836 rows x 7 columns]

----

###  Задание 8.1 ###

Создать в таблице новый признак `year_release`, который соответствует году выпуска фильма.

У скольких фильмов не указан год их выпуска?

***Ответ - 18***

In [69]:
df_sample['year_release'] = df_sample['title'].apply(get_year_release)
display(df_sample.info())
print('Without year:', df_sample.shape[0] - df_sample['year_release'].notna().sum())
print('Without year:', df_sample['year_release'].isna().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Unnamed: 0    100836 non-null  int64  
 1   userId        100836 non-null  int64  
 2   movieId       100836 non-null  int64  
 3   rating        100836 non-null  float64
 4   date          100836 non-null  object 
 5   title         100836 non-null  object 
 6   genres        100836 non-null  object 
 7   year_release  100818 non-null  float64
dtypes: float64(2), int64(3), object(3)
memory usage: 6.2+ MB


None

Without year: 18
Without year: 18


----

###  Задание 8.2 ###

Какой фильм, выпущенный в 1999 году, получил наименьшую среднюю оценку зрителей?

В качестве ответа записать название этого фильма без указания года его выпуска.

***Ответ - Bloodsport: The Dark Kumite***

In [70]:
rates99 = df_sample[df_sample['year_release'] == 1999]
rates99.groupby(by='title')['rating'].agg(['count', 'sum', 'mean']).sort_values(by='mean')

count  sum  mean
title                                                       
Bloodsport: The Dark Kumite (1999)              1  0.5   0.5
Simon Sez (1999)                                1  1.0   1.0
Chill Factor (1999)                             1  1.0   1.0
Source, The (1999)                              1  1.0   1.0
Trippin' (1999)                                 1  1.0   1.0
...                                           ...  ...   ...
Trailer Park Boys (1999)                        1  5.0   5.0
Larry David: Curb Your Enthusiasm (1999)        1  5.0   5.0
Sun Alley (Sonnenallee) (1999)                  1  5.0   5.0
George Carlin: You Are All Diseased (1999)      1  5.0   5.0
Five Senses, The (1999)                         1  5.0   5.0

[261 rows x 3 columns]

----

###  Задание 8.3 ###

Какое сочетание жанров фильмов (`genres`), выпущенных в 2010 году, получило наименьшую среднюю оценку (rating)?

Записать сочетание так же, как оно указано в таблице (через разделитель |, без пробелов).

***Ответ - Action|Sci-Fi***

In [71]:
rates10 = df_sample[df_sample['year_release'] == 2010]
rates10.groupby(by='genres')['rating'].agg(['count', 'sum', 'mean']).sort_values(by='mean')

count   sum      mean
genres                                                  
Action|Sci-Fi                          1   1.0  1.000000
Action|Adventure|Horror                1   1.5  1.500000
Action|Drama|Fantasy                   1   1.5  1.500000
Crime|Romance                          1   1.5  1.500000
Adventure|Comedy|Fantasy               6  11.0  1.833333
...                                  ...   ...       ...
Crime                                  2   9.5  4.750000
Comedy|Musical                         1   5.0  5.000000
Animation|Drama|Fantasy|Mystery        1   5.0  5.000000
Adventure|Children|Comedy|Mystery      1   5.0  5.000000
Animation|Children|Mystery             1   5.0  5.000000

[119 rows x 3 columns]

----

###  Задание 8.4 ###
Какой пользователь (`userId`) посмотрел наибольшее количество различных (уникальных) комбинаций жанров (genres) фильмов?

В качестве ответа записать идентификатор этого пользователя.

***Ответ - 599***

In [72]:
df_sample.groupby(by='userId')['genres'].agg(['count', 'nunique']).sort_values(by='nunique', ascending=False)

count  nunique
userId                
599      2478      524
414      2698      482
448      1864      403
380      1218      399
474      2108      395
...       ...      ...
578        27       15
12         32       15
85         34       13
214        22       13
245        21       13

[610 rows x 2 columns]

----

###  Задание 8.5 ###

Найти пользователя, который выставил **наименьшее** количество оценок, но его средняя оценка фильмам **наибольшая**.

В качестве ответа указать идентификатор этого пользователя.

Чтобы рассчитать несколько параметров для каждого пользователя (количество оценок и среднюю оценку), можно воспользоваться методом `agg()` на сгруппированных данных.

***Ответ - 53***

In [73]:
df_sample.groupby(by='userId')['rating'].agg(['count', 'mean']).sort_values(by=['count', 'mean'], ascending=[True, False])

count      mean
userId                 
53         20  5.000000
595        20  4.200000
189        20  4.100000
569        20  4.000000
278        20  3.875000
...       ...       ...
274      1346  3.235884
448      1864  2.847371
474      2108  3.398956
599      2478  2.642050
414      2698  3.391957

[610 rows x 2 columns]

----

###  Задание 8.6 ###

Найти сочетание жанров (`genres`) за 2018 год, которое имеет наибольший средний рейтинг (среднее по столбцу `rating`), и при этом число выставленных ему оценок (количество значений в столбце `rating`) больше 10.

Записать сочетание так же, как оно указано в таблице (через разделитель |, без пробелов).

***Ответ - Action|Adventure|Sci-Fi***

In [83]:
rates18 = df_sample[df_sample['year_release'] == 2018]
rates18.groupby(by='genres')['rating'].agg(['count', 'mean']).sort_values(by=['count', 'mean'], ascending=[False, False])

count      mean
genres                                                 
Action|Adventure|Sci-Fi                    14  3.928571
Action|Comedy|Sci-Fi                       12  3.875000
Comedy                                      8  3.312500
Action|Adventure|Children|Sci-Fi            5  3.900000
Animation|Comedy                            5  3.500000
Action|Adventure|Animation|Children         4  3.000000
Drama|Horror|Thriller                       4  2.750000
Action|Adventure|Fantasy                    4  2.500000
Adventure|Mystery|Sci-Fi|Thriller           3  3.833333
Action|Adventure|Comedy|Fantasy|Sci-Fi      3  3.666667
Documentary                                 2  4.000000
Action|Adventure|Thriller                   2  3.750000
Action|Adventure|Drama|Sci-Fi|Thriller      2  3.250000
Comedy|Drama                                2  3.250000
Action|Fantasy|Sci-Fi                       2  2.750000
Horror|Mystery|Sci-Fi|Thriller              2  2.250000
Adventure|Comedy|Fantasy                    1  4.500000
Adventure|Thriller                          1  4.500000
Comedy|Fantasy|Sci-Fi                       1  4.500000
Comedy|Romance                              1  4.500000
Action|Comedy|Crime|Horror                  1  4.000000
Action|Crime|Drama|Thriller                 1  4.000000
Action|Animation                            1  3.500000
Action|Mystery|Sci-Fi|Thriller              1  3.500000
Crime|Drama                                 1  3.500000
Horror|Mystery|Thriller                     1  3.500000
Sci-Fi|Thriller                             1  3.500000
Action|Comedy                               1  3.000000
Adventure|Children|Fantasy|Sci-Fi           1  3.000000
Action|Crime|Thriller                       1  2.500000
Comedy|Crime|Drama                          1  2.500000
Drama|Fantasy|Thriller                      1  2.000000
Crime|Drama|Mystery|Thriller                1  1.500000

----

###  Задание 8.7 ###

Добавить в таблицу новый признак `year_rating` - год выставления оценки. Создать сводную таблицу, которая иллюстрирует зависимость среднего рейтинга фильма от года выставления оценки и жанра. Выберите верные варианты ответа, исходя из построенной таблицы

***Ответы***
- **True** За весь период (с 1996 по 2018 год) сочетание жанров Action|Adventure ни разу не получало среднюю оценку ниже 3.
- **False** Наилучшую оценку жанр Action|Adventure|Animation|Children|Comedy|IMAX получил в 2010 году.
- **True** Среди сочетаний жанров, получивших наивысшую среднюю оценку в 2018 году, есть сочетание Animation|Children|Mystery.
- **False** Для жанра Comedy прослеживается тенденция падения рейтинга с каждым годом (с 1996 по 2018).

In [89]:
def get_year_rating(arg):
    candidates = re.findall(r'\d{4}', arg)
    if len(candidates) > 0:
        year = candidates[0]
        return int(year)
    else:
        return None

df_sample['year_rating'] = df_sample['date'].apply(get_year_rating)
display(df_sample)

Unnamed: 0  userId  movieId  rating                 date  \
0                0       1        1     4.0  2000-07-30 18:45:03   
1                1       1        3     4.0  2000-07-30 18:20:47   
2                2       1        6     4.0  2000-07-30 18:37:04   
3                3       1       47     5.0  2000-07-30 19:03:35   
4                4       1       50     5.0  2000-07-30 18:48:51   
...            ...     ...      ...     ...                  ...   
100831      100831     610   166534     4.0  2017-05-03 21:53:22   
100832      100832     610   168248     5.0  2017-05-03 22:21:31   
100833      100833     610   168250     5.0  2017-05-08 19:50:47   
100834      100834     610   168252     5.0  2017-05-03 21:19:12   
100835      100835     610   170875     3.0  2017-05-03 21:20:15   

                                 title  \
0                     Toy Story (1995)   
1              Grumpier Old Men (1995)   
2                          Heat (1995)   
3          Seven (a.k.a. Se7en) (1995)   
4           Usual Suspects, The (1995)   
...                                ...   
100831                    Split (2017)   
100832   John Wick: Chapter Two (2017)   
100833                  Get Out (2017)   
100834                    Logan (2017)   
100835  The Fate of the Furious (2017)   

                                             genres  year_release  year_rating  
0       Adventure|Animation|Children|Comedy|Fantasy        1995.0         2000  
1                                    Comedy|Romance        1995.0         2000  
2                             Action|Crime|Thriller        1995.0         2000  
3                                  Mystery|Thriller        1995.0         2000  
4                            Crime|Mystery|Thriller        1995.0         2000  
...                                             ...           ...          ...  
100831                        Drama|Horror|Thriller        2017.0         2017  
100832                        Action|Crime|Thriller        2017.0         2017  
100833                                       Horror        2017.0         2017  
100834                                Action|Sci-Fi        2017.0         2017  
100835                  Action|Crime|Drama|Thriller        2017.0         2017  

[100836 rows x 9 columns]

In [98]:
# last_df = df_sample.groupby(['genres', 'year_rating'])['rating'].mean().unstack()
# display(last_df)
last_pivot = df_sample.pivot_table(
    values='rating',
    index='genres',
    columns='year_rating',
    aggfunc='mean',
    fill_value=0
)
display(last_pivot)

year_rating                              1996      1997  1998  1999      2000  \
genres                                                                          
(no genres listed)                   0.000000  0.000000   0.0   0.0  0.000000   
Action                               2.730769  3.538462   0.0   0.0  2.588235   
Action|Adventure                     3.454545  4.150000   4.2   4.0  3.738462   
Action|Adventure|Animation           0.000000  0.000000   0.0   0.0  0.000000   
Action|Adventure|Animation|Children  0.000000  0.000000   0.0   0.0  0.000000   
...                                       ...       ...   ...   ...       ...   
Sci-Fi|Thriller                      2.666667  3.400000   0.0   4.0  2.142857   
Sci-Fi|Thriller|IMAX                 0.000000  0.000000   0.0   0.0  0.000000   
Thriller                             3.838095  3.923077   3.8   3.7  3.087912   
War                                  0.000000  0.000000   0.0   4.5  3.000000   
Western                              3.117647  3.000000   0.0   4.0  4.058824   

year_rating                              2001      2002      2003      2004  \
genres                                                                        
(no genres listed)                   0.000000  0.000000  0.000000  0.000000   
Action                               3.000000  2.750000  3.833333  2.700000   
Action|Adventure                     3.500000  4.304348  3.277778  4.136364   
Action|Adventure|Animation           0.000000  0.000000  0.000000  0.000000   
Action|Adventure|Animation|Children  0.000000  0.000000  0.000000  0.000000   
...                                       ...       ...       ...       ...   
Sci-Fi|Thriller                      2.500000  3.600000  3.142857  0.000000   
Sci-Fi|Thriller|IMAX                 0.000000  0.000000  0.000000  0.000000   
Thriller                             3.477273  3.583333  3.250000  3.464286   
War                                  3.000000  3.500000  3.000000  3.000000   
Western                              3.111111  3.000000  4.000000  3.800000   

year_rating                              2005  ...      2009      2010  \
genres                                         ...                       
(no genres listed)                   0.000000  ...  0.000000  0.000000   
Action                               3.357143  ...  0.000000  3.500000   
Action|Adventure                     3.413043  ...  3.714286  3.437500   
Action|Adventure|Animation           0.000000  ...  0.000000  0.000000   
Action|Adventure|Animation|Children  0.000000  ...  0.000000  0.000000   
...                                       ...  ...       ...       ...   
Sci-Fi|Thriller                      2.750000  ...  4.000000  2.500000   
Sci-Fi|Thriller|IMAX                 0.000000  ...  0.000000  0.000000   
Thriller                             3.411765  ...  2.964286  2.833333   
War                                  0.000000  ...  4.000000  0.000000   
Western                              4.500000  ...  3.375000  3.666667   

year_rating                              2011      2012      2013   2014  \
genres                                                                     
(no genres listed)                   0.000000  0.000000  0.000000  0.000   
Action                               4.500000  2.625000  2.500000  0.000   
Action|Adventure                     3.888889  3.900000  3.833333  3.875   
Action|Adventure|Animation           0.000000  0.000000  0.000000  3.000   
Action|Adventure|Animation|Children  0.000000  4.000000  3.000000  3.500   
...                                       ...       ...       ...    ...   
Sci-Fi|Thriller                      1.833333  3.500000  3.833333  0.000   
Sci-Fi|Thriller|IMAX                 4.000000  3.666667  4.000000  4.000   
Thriller                             3.625000  3.083333  3.666667  4.000   
War                                  0.000000  0.000000  0.000000  0.000   
Western                              4.0000

In [103]:
last_pivot.loc['Action|Adventure', :]

year_rating
1996    3.454545
1997    4.150000
1998    4.200000
1999    4.000000
2000    3.738462
2001    3.500000
2002    4.304348
2003    3.277778
2004    4.136364
2005    3.413043
2006    4.090909
2007    3.666667
2008    3.656250
2009    3.714286
2010    3.437500
2011    3.888889
2012    3.900000
2013    3.833333
2014    3.875000
2015    3.560976
2016    3.722222
2017    3.527778
2018    3.397436
Name: Action|Adventure, dtype: float64

In [107]:
last_pivot.loc["Action|Adventure|Animation|Children|Comedy|IMAX", :].sort_values(ascending=False)

year_rating
2013    5.000000
2018    3.700000
2016    3.666667
2012    3.550000
2014    3.500000
2010    3.500000
2015    3.375000
2009    3.250000
2017    1.000000
2008    0.000000
2011    0.000000
1996    0.000000
1997    0.000000
2006    0.000000
2005    0.000000
2004    0.000000
2003    0.000000
2002    0.000000
2001    0.000000
2000    0.000000
1999    0.000000
1998    0.000000
2007    0.000000
Name: Action|Adventure|Animation|Children|Comedy|IMAX, dtype: float64

In [115]:
last_pivot.loc[:, 2018].sort_values(ascending=False).head(23)

genres
Comedy|Horror|Mystery                        5.00
Comedy|Fantasy|Horror|Musical|Thriller       5.00
Animation|Children|Mystery                   5.00
Comedy|Crime|Horror|Thriller                 5.00
Drama|Horror|Mystery|Sci-Fi|Thriller         5.00
Adventure|Children|Comedy|Fantasy|Mystery    5.00
Action|Crime|Thriller|Western                5.00
Comedy|Crime|Fantasy                         5.00
Crime|Mystery                                5.00
Drama|Romance|War                            5.00
Animation|Comedy|Drama                       5.00
Animation|Comedy|Drama|Fantasy|Sci-Fi        5.00
Action|Crime|Sci-Fi|Thriller                 5.00
Action|Comedy|Documentary                    5.00
Comedy|Drama|Fantasy|Mystery|Romance         5.00
Mystery|Romance|Sci-Fi|Thriller              5.00
Romance|Sci-Fi                               5.00
Animation|Crime|Drama                        5.00
Crime|Fantasy|Horror                         5.00
Adventure|Fantasy|Musical                  

In [116]:
last_pivot.loc['Comedy', :]

year_rating
1996    3.228571
1997    3.409091
1998    3.000000
1999    3.606061
2000    3.141291
2001    3.318408
2002    3.198556
2003    3.120066
2004    3.356877
2005    2.963325
2006    3.347534
2007    2.928187
2008    3.426667
2009    3.273292
2010    3.179825
2011    3.232877
2012    3.478477
2013    3.489474
2014    3.271429
2015    3.106183
2016    3.373431
2017    2.852668
2018    3.121296
Name: Comedy, dtype: float64

## Новые датасеты ##

Дано два файла:
- `orders.csv`, содержащий данные о заказах;
- `products.csv`, содержащий данные о товарах.

Объединить заданные таблицы в таблицу `orders_products`, чтобы в результирующей таблице оказалась информация обо всех заказах, но не оказалось информации о продуктах, на которых заказов ещё не поступало.

In [129]:
orders = pd.read_csv('data/orders.csv', sep=';')
products = pd.read_csv('data/products.csv', sep=';')
# display(orders.info())
# display(products.info())
display(orders)
display(products)

Дата создания  Order ID  ID Покупателя  \
0   09.11.2019 21:55:51         9             10   
1   09.11.2019 15:05:57         8              9   
2   09.11.2019 15:05:57         8              9   
3   09.11.2019 12:50:07         7              8   
4   09.11.2019 12:00:00         6              1   
5   09.11.2019 12:00:00         6              1   
6   08.11.2019 08:36:22         5              5   
7   08.11.2019 08:36:22         4              9   
8   08.11.2019 08:36:22         3              8   
9   08.11.2019 08:36:22         3              8   
10  08.11.2019 08:36:21         2              7   
11  08.11.2019 08:36:21         2              7   
12  08.11.2019 08:36:21         2              7   
13  08.11.2019 08:36:21         2              7   
14  08.11.2019 08:36:20         1              5   
15  08.11.2019 08:36:20         1              5   
16  08.11.2019 08:36:20         1              5   
17  01.01.2001 00:00:00         0              1   

                             Статус Оплачен Отменен Отгружен  ID товара  \
0          Принят, ожидается оплата     Нет     Нет      Нет        103   
1          Принят, ожидается оплата     Нет     Нет      Нет         86   
2          Принят, ожидается оплата     Нет     Нет      Нет        104   
3          Принят, ожидается оплата     Нет     Нет      Нет        104   
4          Принят, ожидается оплата     Нет     Нет      Нет        104   
5          Принят, ожидается оплата     Нет     Нет      Нет        103   
6                           Отменён     Нет      Да      Нет        124   
7          Принят, ожидается оплата     Нет     Нет       Да         91   
8   Оплачен, формируется к отправке      Да     Нет      Нет        103   
9   Оплачен, формируется к отправке      Да     Нет      Нет        104   
10                         Выполнен      Да     Нет       Да         74   
11                         Выполнен      Да     Нет       Да         53   
12                         Выполнен      Да     Нет       Да        137   
13                         Выполнен      Да     Нет       Да         47   
14  Оплачен, формируется к отправке      Да     Нет      Нет         71   
15  Оплачен, формируется к отправке      Да     Нет      Нет         86   
16  Оплачен, формируется к отправке      Да     Нет      Нет         51   
17  Оплачен, формируется к отправке      Да     Нет      Нет        666   

    Количество  
0            5  
1          100  
2           10  
3            7  
4            5  
5            5  
6            1  
7            1  
8            3  
9            3  
10           1  
11           1  
12           1  
13           1  
14           1  
15           1  
16           1  
17           1

Product_ID                                   Name  Price CURRENCY
0           47                   Шатны Полосатый рейс   2999      RUR
1           51               Платье Аленький цветочек   4999      RUR
2           53                 Штаны Цветочная Поляна   4999      RUR
3           71                    Платье Ночная Жизнь   7999      RUR
4           74               Платье Ночная Жизнь XXXL   8999      RUR
5           86                     Носки Простые, муж     45      RUR
6           91                     Носки Честные, муж     50      RUR
7          103                  Носки Подарочные, муж    199      RUR
8          104                  Носки Подарочные, жен    249      RUR
9          124                   Носки беговые Camino    999      RUR
10         137              Гольфы детские Снегурочка     99      RUR
11         138               Гольфы детские Питер Пен     99      RUR
12         139                       Сумка для носков    500      RUR
13         140  Товар, которому не придумано название      1      RUR

In [132]:
orders_products = orders.sort_values(by=['Order ID', 'ID товара']).merge(
    products,
    how='left',
    left_on='ID товара',
    right_on='Product_ID'
)
display(orders_products)

Дата создания  Order ID  ID Покупателя  \
0   01.01.2001 00:00:00         0              1   
1   08.11.2019 08:36:20         1              5   
2   08.11.2019 08:36:20         1              5   
3   08.11.2019 08:36:20         1              5   
4   08.11.2019 08:36:21         2              7   
5   08.11.2019 08:36:21         2              7   
6   08.11.2019 08:36:21         2              7   
7   08.11.2019 08:36:21         2              7   
8   08.11.2019 08:36:22         3              8   
9   08.11.2019 08:36:22         3              8   
10  08.11.2019 08:36:22         4              9   
11  08.11.2019 08:36:22         5              5   
12  09.11.2019 12:00:00         6              1   
13  09.11.2019 12:00:00         6              1   
14  09.11.2019 12:50:07         7              8   
15  09.11.2019 15:05:57         8              9   
16  09.11.2019 15:05:57         8              9   
17  09.11.2019 21:55:51         9             10   

                             Статус Оплачен Отменен Отгружен  ID товара  \
0   Оплачен, формируется к отправке      Да     Нет      Нет        666   
1   Оплачен, формируется к отправке      Да     Нет      Нет         51   
2   Оплачен, формируется к отправке      Да     Нет      Нет         71   
3   Оплачен, формируется к отправке      Да     Нет      Нет         86   
4                          Выполнен      Да     Нет       Да         47   
5                          Выполнен      Да     Нет       Да         53   
6                          Выполнен      Да     Нет       Да         74   
7                          Выполнен      Да     Нет       Да        137   
8   Оплачен, формируется к отправке      Да     Нет      Нет        103   
9   Оплачен, формируется к отправке      Да     Нет      Нет        104   
10         Принят, ожидается оплата     Нет     Нет       Да         91   
11                          Отменён     Нет      Да      Нет        124   
12         Принят, ожидается оплата     Нет     Нет      Нет        103   
13         Принят, ожидается оплата     Нет     Нет      Нет        104   
14         Принят, ожидается оплата     Нет     Нет      Нет        104   
15         Принят, ожидается оплата     Нет     Нет      Нет         86   
16         Принят, ожидается оплата     Нет     Нет      Нет        104   
17         Принят, ожидается оплата     Нет     Нет      Нет        103   

    Количество  Product_ID                       Name   Price CURRENCY  
0            1         NaN                        NaN     NaN      NaN  
1            1        51.0   Платье Аленький цветочек  4999.0      RUR  
2            1        71.0        Платье Ночная Жизнь  7999.0      RUR  
3            1        86.0         Носки Простые, муж    45.0      RUR  
4            1        47.0       Шатны Полосатый рейс  2999.0      RUR  
5            1        53.0     Штаны Цветочная Поляна  4999.0      RUR  
6            1        74.0   Платье Ночная Жизнь XXXL  8999.0      RUR  
7            1       137.0  Гольфы детские Снегурочка    99.0      RUR  
8            3       103.0      Носки Подарочные, муж   199.0      RUR  
9            3       104.0      Носки Подарочные, жен   249.0      RUR  
10           1        91.0         Носки Честные, муж    50.0      RUR  
11           1       124.0       Носки беговые Camino   999.0      RUR  
12           5       103.0      Носки Подарочные, муж   199.0      RUR  
13           5       104.0      Носки Подарочные, жен   249.0      RUR  
14           7       104.0      Носки Подарочные, жен   249.0      RUR  
15         100        86.0         Носки Простые, муж    45.0      RUR  
16          10       104.0      Носки Подарочные, жен   249.0      RUR  
17           5       103.0      Носки Подарочные, муж   199.0      RUR

----

###  Задание 8.8 ###

Какой идентификатор (`Order ID`) имеет заказ, для которого не оказалось информации о товаре?

***Ответ - 0***

----

###  Задание 8.9 ###

На какой товар была произведена отмена?

В качестве ответа записать название этого товара (Name).

***Ответ - Носки беговые Camino***

----

###  Задание 8.10 ###

Какой покупатель принёс наибольшую суммарную прибыль интернет-магазину за указанный период?

В ответ записать идентификатор этого покупателя (`ID Покупателя`).

Прибыль состоит **только из оплаченных заказов** и рассчитывается как **количество купленного товара, умноженное на его цену**.

***Ответ - 7***

In [134]:
mask1 = orders_products['Статус'] == 'Оплачен, формируется к отправке'
mask2 = orders_products['Статус'] == 'Выполнен'
profits = orders_products[mask1 | mask2]
display(profits)

Дата создания  Order ID  ID Покупателя  \
0  01.01.2001 00:00:00         0              1   
1  08.11.2019 08:36:20         1              5   
2  08.11.2019 08:36:20         1              5   
3  08.11.2019 08:36:20         1              5   
4  08.11.2019 08:36:21         2              7   
5  08.11.2019 08:36:21         2              7   
6  08.11.2019 08:36:21         2              7   
7  08.11.2019 08:36:21         2              7   
8  08.11.2019 08:36:22         3              8   
9  08.11.2019 08:36:22         3              8   

                            Статус Оплачен Отменен Отгружен  ID товара  \
0  Оплачен, формируется к отправке      Да     Нет      Нет        666   
1  Оплачен, формируется к отправке      Да     Нет      Нет         51   
2  Оплачен, формируется к отправке      Да     Нет      Нет         71   
3  Оплачен, формируется к отправке      Да     Нет      Нет         86   
4                         Выполнен      Да     Нет       Да         47   
5                         Выполнен      Да     Нет       Да         53   
6                         Выполнен      Да     Нет       Да         74   
7                         Выполнен      Да     Нет       Да        137   
8  Оплачен, формируется к отправке      Да     Нет      Нет        103   
9  Оплачен, формируется к отправке      Да     Нет      Нет        104   

   Количество  Product_ID                       Name   Price CURRENCY  
0           1         NaN                        NaN     NaN      NaN  
1           1        51.0   Платье Аленький цветочек  4999.0      RUR  
2           1        71.0        Платье Ночная Жизнь  7999.0      RUR  
3           1        86.0         Носки Простые, муж    45.0      RUR  
4           1        47.0       Шатны Полосатый рейс  2999.0      RUR  
5           1        53.0     Штаны Цветочная Поляна  4999.0      RUR  
6           1        74.0   Платье Ночная Жизнь XXXL  8999.0      RUR  
7           1       137.0  Гольфы детские Снегурочка    99.0      RUR  
8           3       103.0      Носки Подарочные, муж   199.0      RUR  
9           3       104.0      Носки Подарочные, жен   249.0      RUR

In [142]:
profits.groupby(by='ID Покупателя')['Price'].sum().sort_values(ascending=False)

ID Покупателя
7    17096.0
5    13043.0
8      448.0
1        0.0
Name: Price, dtype: float64